ledger/eval/txntracer.go

```diff
@@ -48,12 +48,31 @@ type StateDeltaSubset struct {
 }
 
 func convertStateDelta(delta ledgercore.StateDelta) StateDeltaSubset {
+	// The StateDelta object returned through the EvalTracer has its values deleted between txn groups to avoid
+	// reallocation during evaluation.
+	// This means the map values need to be copied (to avoid deletion) since they are all passed by reference.
+	kvmods := make(map[string]ledgercore.KvValueDelta)
+	for k1, v1 := range delta.KvMods {
+		kvmods[k1] = v1
+	}
+	txids := make(map[transactions.Txid]ledgercore.IncludedTransactions)
+	for k2, v2 := range delta.Txids {
+		txids[k2] = v2
+	}
+	txleases := make(map[ledgercore.Txlease]basics.Round)
+	for k3, v3 := range delta.Txleases {
+		txleases[k3] = v3
+	}
+	creatables := make(map[basics.CreatableIndex]ledgercore.ModifiedCreatable)
+	for k4, v4 := range delta.Creatables {
+		creatables[k4] = v4
+	}
 	return StateDeltaSubset{
 		Accts:      delta.Accts,
```

Comment:
`delta.Accts` is also subject to recycling. I suspect you're not seeing a difference in your test because the `reset` function is not immediately destructive, but the underlying memory is still reserved for the next user of the cow.

https://github.com/algorand/go-algorand/blob/b988971925f5c9e7399c0df0393bb99f7376b1a6/ledger/ledgercore/statedelta.go#L258-L260

Reply:
This is tricky. It seems like you're right so I've copied all of the data out of the AccountDeltas as well, but it is hard for me to reason about which CoWs are being reused across evaluation of multiple txn groups. 

I even added a second group in the same block and it didn't affect the returned results.

Reply:
Before your fix in bdf35d8, when I run this test with `-count=100`, I see some errors, so it's definitely an issue.

Strangely adding the `-race` flag doesn't seem to catch this, though it has caught similar mistakes I've made around cow recycling.

The fix looks good and I don't see any more errors when testing locally.